### PR TITLE
wicked: Add legacy macvtap test case

### PR DESCRIPTION
**Pull request by asmorodskyi**
Add legacy macvtap test case

### CODE REVIEW

**Comment by pevik** [comment]
@asmorodskyi I wonder if it was intentional to add C binary (`data/wicked/check_macvtap`). Having SDK , install gcc and compile the source would be IMHO much better.

**Comment by asmorodskyi** [comment]
install additional package and compile source code would add additional reasons to fail with issue not related to wicked. This test currently suppose to run on single architecture ( x86 ) so I don't see obvious reasons why it is "much better" can you elaborate ?

**Comment by pevik** [comment]
I'd check for argc (not nice to get segfault when run without parameters), but understand you don't care.
And I wonder, if it'd be achieved with perl only code (I mean plain perl, not perlxs).

**Comment by asmorodskyi** [comment]
you right I think it this not necessary the same as checking that we getting really IPs and not some random strings. This binary designed for single test suite so there is no chance to do something wrong.
I am using standard `assert_script_run` to execute it so perlxs is not relevant and not needed IMHO

**Comment by pevik** [comment]
perlxs: You don't understand me: I thought using perl (or python or any scripting language), so you don't have to include compiled binary (which is ugly) or employ `gcc`. It looks to me that you just manipulating with binary data in `int[]`, which is not limited only for C :).
BTW you're mixing tags with spaces in already existing files (at least at `{ref,sut}/t15_macvtap_legacy.pm`), but that will be visible in Mr. Travis :).

**Comment by asmorodskyi** [comment]
Finally all are working as expected :) 
Prove runs : 
http://kimball.arch.suse.de/tests/631
http://kimball.arch.suse.de/tests/630

**Comment by asmorodskyi** [comment]
@mtomaschewski , @rtorrero can you please take a look ?

**Comment by jlausuch** [comment]
LGTM but would be nice to have the +1 from the network experts

**Comment by asmorodskyi** [comment]
yeah also prefer not to merge it and give some time to Marius and/or Ruben to review it ...

**Comment by mtomaschewski** [comment]
About compiling.. wicked-testsuite were compiling it in obs/ibs for each image in the wicked-testsuite package. Using plain perl would be an option and does not need compilation at all and as this can be
done in plain perl too ... the most "difficult" part here is to put the expected IPs (as ntohs) into the "expected" data buffer, the rest is basically byte received-data vs. expected-data compare + file read.
But effectively, the C version existed and as long as it works...

**Comment by asmorodskyi** [comment]
removed binary , start using gcc as many people suggesting me . verification runs :
- http://kimball.arch.suse.de/tests/654
- http://kimball.arch.suse.de/tests/653

File: data/wicked/check_macvtap.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
+// Small program to test macvtap interfaces
+// It assumes that
+//                         MAC        0E:0E:0E:0E:0E:0E
```
Comment (pevik) [inline data/wicked/check_macvtap.c]: I'd remove that huge whitespace at the front, but that's very minor.

File: data/wicked/check_macvtap.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+    if (differ == 1)
+    {
+        exit(-4);
+    }
+    else
+    {
+        printf("Success listening to tap device %s, received the expected ARP packet\n", fn);
+    }
```
Comment (pevik) [inline data/wicked/check_macvtap.c]: I'd prefer using brackets on the single line as `if`/`else` as it's more compact (even for code which does need brackets at all), but of course your code => your coding style :)

File: data/wicked/check_macvtap.c [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+// Main program
+int main(int argc, char **argv)
+{
+    if(argc < 2) {
```
Comment (pevik) [inline data/wicked/check_macvtap.c]: `if` without space before.

File: data/wicked/check_macvtap.c [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        printf("No input parameters. Exit \n");
+        exit(-5);
+    }
+    // getting source and destination ips from command line and inserting them into
+    // dedicated places in packet
+    struct in_addr *source_ip = (struct in_addr *)&arp_packet[38];
+    inet_pton(AF_INET, argv[1], source_ip);
+    struct in_addr *dest_ip = (struct in_addr *)&arp_packet[48];
```
Comment (pevik) [inline data/wicked/check_macvtap.c]: Can these magic `38` and `48` sizes be described with some `#define` ?

File: tests/wicked/advanced/ref/t15_macvtap_legacy.pm [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+use strict;
+use warnings;
+use testapi;
+
+
+sub run {
+    my ($self) = @_;
+    record_info('Info', 'Create a macvtap interface from legacy ifcfg files');
+
+}
```
Comment (pevik) [inline tests/wicked/advanced/ref/t15_macvtap_legacy.pm]: Why this empty line?

File: tests/wicked/advanced/sut/t15_macvtap_legacy.pm [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+our $macvtap_log = '/tmp/macvtap_results.txt';
+
+sub run {
+    my ($self) = @_;
+    record_info('Info', 'Create a macvtap interface from legacy ifcfg files');
+    my $config     = '/etc/sysconfig/network/ifcfg-macvtap1';
+    my $iface      = iface();
+    my $ip_address = $self->get_ip(type => 'macvtap', netmask => 1);
+    $ip_address =~ s'/'\\/';
```
Comment (pevik) [inline tests/wicked/advanced/sut/t15_macvtap_legacy.pm]: Maybe worth of adding support to `get_ip` for returning escaped `/` via parameter.

File: tests/wicked/advanced/sut/t15_macvtap_legacy.pm [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    my $ref_ip = $self->get_ip(type => 'host', netmask => 0, is_wicked_ref => 1);
+    $self->get_from_data('wicked/ifcfg/macvtap1',    $config);
+    $self->get_from_data('wicked/ifcfg/macvtap_eth', '/etc/sysconfig/network/ifcfg-' . $iface);
+    $self->get_from_data('wicked/check_macvtap.c',   'check_macvtap.c', executable => 1);
+    zypper_call('in gcc');
+    assert_script_run('gcc ./check_macvtap.c -o check_macvtap');
+    script_run('chmod +x ./check_macvtap');
+    assert_script_run("sed 's/iface/$iface/' -i $config");
```
Comment (pevik) [inline tests/wicked/advanced/sut/t15_macvtap_legacy.pm]: `./` not needed in these two (of course not a problem).

File: data/wicked/check_macvtap.c (hunk 4 above)
Comment (pevik) [reply data/wicked/check_macvtap.c]: \+ the same applies for `char fn[16]` (some size defined and deliver `15` from it).

File: data/wicked/check_macvtap.c (hunk 2 above)
Comment (asmorodskyi) [reply data/wicked/check_macvtap.c]: totally fine for me , but with one minor correction from my side don't want to do it line by line over and over again . do we have some "tidy" for C which I  could run against this file to cover all your standards regarding code style ? :)

File: data/wicked/check_macvtap.c (hunk 3 above)
Comment (asmorodskyi) [reply data/wicked/check_macvtap.c]: read my answer above

File: tests/wicked/advanced/sut/t15_macvtap_legacy.pm (hunk 7 above)
Comment (asmorodskyi) [reply tests/wicked/advanced/sut/t15_macvtap_legacy.pm]: I prefer to have it

File: tests/wicked/advanced/sut/t15_macvtap_legacy.pm (hunk 6 above)
Comment (asmorodskyi) [reply tests/wicked/advanced/sut/t15_macvtap_legacy.pm]: maybe but for me it looks like a separate PR . this code pattern ( get_ip call and escaping `/` )  used in few other places so I would need to touch to many thing not related to introduction of macvtap test and this is what I am trying to achieve with this PR

File: tests/wicked/advanced/ref/t15_macvtap_legacy.pm (hunk 5 above)
Comment (asmorodskyi) [reply tests/wicked/advanced/ref/t15_macvtap_legacy.pm]: fixed

**Comment by jlausuch** [comment]
LGTM

**Pull request merged**